graph/stream-enact-events.go

```diff
@@ -82,43 +82,86 @@ func NewVertex[V VPI[V], E EPI[E], M MVI[M], N any, A Algorithm[V, E, M, N]](alg
 		}
 
 		pos := ^uint32(0)
-		if event.EventType() == ADD {
-			vs := gt.VertexStructure(didx)
-			pos = vs.InEventPos
-			if onInEdgeAddFunc != nil {
-				onInEdgeAddFunc(g, gt, gt.Vertex(didx), gt.VertexProperty(didx), didx, pos, &event)
+		if gt.InputEvent.EventType() == ADD {
+			vs := gt.VertexStructure(vidx)
+			if gt.InputEvent.EventIdx()%2 == 1 { // We are the original source, they are the original destination.
+				if g.Options.Undirected { // For undirected, we make the edge both ways, so we are also a destination.
+					pos = vs.InEventPos
+					if onInEdgeAddFunc != nil {
+						onInEdgeAddFunc(g, gt, gt.Vertex(vidx), gt.VertexProperty(vidx), vidx, pos, &gt.InputEvent)
+					}
+					vs.InEventPos++
+				}
+			} else { // They are the original source, we are the original destination.
+				pos = vs.InEventPos
+				if onInEdgeAddFunc != nil {
+					onInEdgeAddFunc(g, gt, gt.Vertex(vidx), gt.VertexProperty(vidx), vidx, pos, &gt.InputEvent)
+				}
+				vs.InEventPos++
 			}
-			vs.InEventPos++
-			gt.NumInEvents++ // Thread total; unused.
 		}
 
-		// Will always succeed (range check is lte current space)
-		gt.ToRemitQueue.Offer(RawEdgeEvent[E]{TypeAndEventIdx: event.TypeAndEventIdx, SrcRaw: event.SrcRaw, DstRaw: event.DstRaw, Edge: Edge[E]{Property: event.EdgeProperty, Didx: didx, Pos: pos}})
+		// Stash the event first.
+		gt.RemitEvent = RemitEvent[E]{TypeAndEventIdx: gt.InputEvent.TypeAndEventIdx, Order: gt.InputEvent.ThreadOrderSrc, Edge: Edge[E]{Property: gt.InputEvent.EdgeProperty, Didx: vidx, Pos: pos}}
+		// If we dont have room, we fail.
+		if (gt.InputEvent.ThreadOrderDst - topologyFlushedUpTo) >= uint64(len(gt.TopologyEventBuff)) {
+			//log.Debug().Msg("T[" + utils.F("%02d", gt.Tidx) + "] Could not insert event, will try again next time ")
+			gt.FailedRemit = true
+			return false, count, largest, true, false
+		}
+		// If we can't remit, we fail.
+		if ok = g.GraphThreads[gt.InputEvent.SrcRaw.Within(g.NumThreads)].FromRemitQueue.PutFast(gt.RemitEvent, gt.InputEvent.ThreadOrderSrc); !ok {
+			//log.Debug().Msg("T[" + utils.F("%02d", gt.Tidx) + "] Could not emplace event, will try again next time ")
+			gt.FailedRemit = true
+			return false, count, largest, true, false
+		}
+		gt.TopologyEventBuff[gt.InputEvent.ThreadOrderDst-topologyFlushedUpTo] = RawEdgeEvent[E]{TypeAndEventIdx: gt.InputEvent.TypeAndEventIdx, Sidx: vidx, Edge: gt.RemitEvent.Edge}
+		largest = gt.InputEvent.ThreadOrderDst
 	}
-	if spaceAvailable != 0 && count == 0 {
+
+	if count == 0 {
 		closed = gt.FromEmitQueue.IsClosed()
 	}
-	return closed, count
+	return closed, count, largest, false, false
 }
 
 // Injects expired edges into the topology event buffer. To be done after the topology event buffer has been remapped with internal source ids.
 func InjectExpired[EP EPP[E], V VPI[V], E EPI[E], M MVI[M], N any](g *Graph[V, E, M, N], gt *GraphThread[V, E, M, N], changeCount uint64, uniqueCount uint64, delOnExpire uint64) (newUniqueCount uint64) {
```

Comment:
Because of the way the TopologyEventBuff is now indexed (via InputEvent ThreadOrder and topologyFlushedUpTo), this will need to be updated to inject deletes in a different way. (They can't go into the main TopologyEventBuff)
One way could be to have the Emitter track them and inject them, but this would be terrible for performance.
Probably better to keep a separate buffer of the expired events and have perhaps a flag in the VertexPendingBuff that signifies which buffer to pull from for the next change...